aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java

```diff
@@ -238,6 +236,68 @@ private Completable initializeStorageAdapter(Context context) {
         ));
     }
 
+    private void waitForInitialization(@NonNull Action onComplete, @NonNull Consumer<DataStoreException> onError) {
+        Completable.create(emitter -> {
+            categoryInitializationsPending.await();
+            emitter.onComplete();
+        })
+                .timeout(LIFECYCLE_TIMEOUT_MS, TimeUnit.MILLISECONDS)
+                .subscribeOn(Schedulers.io())
+                .subscribe(
+                        () -> onComplete.call(),
+                        throwable -> onError.accept(new DataStoreException("Request failed because DataStore is not " +
+                                "initialized.", throwable, "Retry your request."))
+                );
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void start(@NonNull Action onComplete, @NonNull Consumer<DataStoreException> onError) {
+        waitForInitialization(() -> {
+            try {
+                orchestrator.start();
+            } catch (DataStoreException exception) {
+                onError.accept(exception);
+                return;
+            }
+            onComplete.call();
```

Comment:
> Could both onComplete and onError fire for the same call? I think onComplete will always fire, but if an exception is thrown by orchestrator.start() the onError(...) callback will also be invoked which could lead to devs assuming that if onComplete() fires, they're OK to move on with the rest of their call flow.

Hmm, did I mess this up?   I think `onComplete` will _not_ be called on error, because there's a `return` on line 263, right?

Reply:
Nevermind. For whatever reason I thought the onComplete.call() was outside of the curly braces.